pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,394 @@
+	machine, err := r.findMachineFromNodeByProviderID(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find machine from node %q by ProviderID: %v", node.GetName(), err)
+	}
+	if machine != nil {
+		return machine, nil
+	}
+
+	machine, err = r.findMachineFromNodeByIP(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by internal IP: %v", machine.GetName(), err)
+	}
+	return machine, nil
+}
+
+func (r *ReconcileNodeLink) findMachineFromNodeByProviderID(node corev1.Node) (*mapiv1beta1.Machine, error) {
+	klog.V(3).Infof("Finding machine from node %q by ProviderID", node.GetName())
+	if node.Spec.ProviderID == "" {
+		klog.Warningf("Node %q has no providerID", node.GetName())
+		return nil, nil
+	}
+
+	machine, ok := r.providerIDToMachine[node.Spec.ProviderID]
+	if ok {
+		klog.V(3).Infof("Found machine %q for node %q with providerID %q from cache", machine.GetName(), node.GetName(), node.Spec.ProviderID)
+		return machine, nil
+	}
+
+	machineList := &mapiv1beta1.MachineList{}
+	if err := r.client.List(
+		context.TODO(),
+		&client.ListOptions{},
+		machineList,
+	); err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+	for i := range machineList.Items {
+		if machineList.Items[i].Spec.ProviderID == nil {
+			continue
+		}
+		if *machineList.Items[i].Spec.ProviderID == node.Spec.ProviderID {
+			klog.V(3).Infof("Found machine %q for node %q with providerID %q", machineList.Items[i].GetName(), node.GetName(), node.Spec.ProviderID)
+			r.providerIDToMachine[node.Spec.ProviderID] = &machineList.Items[i]
+			return r.providerIDToMachine[node.Spec.ProviderID], nil
+		}
+	}
+	return nil, nil
+}
+
+func (r *ReconcileNodeLink) findMachineFromNodeByIP(node corev1.Node) (*mapiv1beta1.Machine, error) {
+	klog.V(3).Infof("Finding machine from node %q by IP", node.GetName())
+	// TODO create hash cached on mapFn and lookup here
+	machineList := &mapiv1beta1.MachineList{}
+	if err := r.client.List(
+		context.TODO(),
+		&client.ListOptions{},
+		machineList,
+	); err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+
+	var nodeInternalAddress string
+	for _, a := range node.Status.Addresses {
+		if a.Type == corev1.NodeInternalIP {
+			nodeInternalAddress = a.Address
+			klog.V(3).Infof("Found internal IP for node %q: %q", node.GetName(), nodeInternalAddress)
+			break
+		}
+	}
+
+	if nodeInternalAddress == "" {
+		klog.Warningf("not found internal IP for node %q", node.GetName())
+		return nil, nil
+	}
+
+	for i := range machineList.Items {
+		for _, a := range machineList.Items[i].Status.Addresses {
+			// Use the internal IP to look for matches:
+			if a.Type == corev1.NodeInternalIP {
+				if a.Address == nodeInternalAddress {
+					klog.V(3).Infof("Found machine %q for node %q with internal IP %q", machineList.Items[i].GetName(), node.GetName(), a.Address)
+					return &machineList.Items[i], nil
+				}
+				break
```

Comment:
Can you drop this `break` here?  See https://github.com/openshift/machine-api-operator/pull/314 which would be dropped if this PR goes in first

Reply:
sure, I'll make sure to cover #314 if this merges first

Reply:
done https://github.com/openshift/machine-api-operator/pull/321/files#diff-69885f1e6e1f6b4cb889ce80021c95a1R82